src/csharp/Microsoft.Spark/Broadcast.cs

```diff
@@ -0,0 +1,243 @@
+        /// <param name="value">Broadcast value to be written to the file</param>
+        private void WriteBroadcastValueToFile(string tempDir, object value)
+        {
+            Directory.CreateDirectory(tempDir);
+            using FileStream f = File.Create(_path);
+            Dump(value, f);
+        }
+
+        /// <summary>
+        /// Function that serializes and stores the object passed to the given Stream.
+        /// </summary>
+        /// <param name="value">Serializable object</param>
+        /// <param name="stream">Stream to which the object is serialized</param>
+        private void Dump(object value, Stream stream)
+        {
+            var formatter = new BinaryFormatter();
+            formatter.Serialize(stream, value);
+        }
+
+        /// <summary>
+        /// Calls the necessary functions to create org.apache.spark.broadcast.Broadcast object
+        /// for Spark versions 2.3.0 and 2.3.1 and returns the JVMObjectReference object.
+        /// </summary>
+        /// <param name="tempDir">Path where file is to be created</param>
+        /// <param name="sparkContext">JVMObjectReference object of SparkContext</param>
+        /// <param name="value">Broadcast value of type object</param>
+        /// <returns></returns>
+        private JvmObjectReference CreateBroadcast_V2_3_1_Below(
+            string tempDir,
+            JvmObjectReference sparkContext,
+            object value)
+        {
+            WriteBroadcastValueToFile(tempDir, value);
+            var javaSparkContext = (JvmObjectReference)sparkContext.Jvm.CallStaticJavaMethod(
+                "org.apache.spark.api.java.JavaSparkContext",
+                "fromSparkContext",
+                sparkContext);
+            return (JvmObjectReference)sparkContext.Jvm.CallStaticJavaMethod(
+                "org.apache.spark.api.python.PythonRDD",
+                "readBroadcastFromFile",
+                javaSparkContext,
+                _path);
+        }
+
+        /// <summary>
+        /// Calls the necessary Spark functions to create org.apache.spark.broadcast.Broadcast
+        /// object for Spark versions 2.3.2 and above, and returns the JVMObjectReference object.
+        /// </summary>
+        /// <param name="sc">SparkContext object</param>
+        /// <param name="tempDir">Path where file is to be created</param>
+        /// <param name="value">Broadcast value of type object</param>
+        /// <returns></returns>
+        private JvmObjectReference CreateBroadcast_V2_3_2_Above(
+            SparkContext sc,
+            string tempDir,
+            object value)
+        {
+            var pythonBroadcast = (JvmObjectReference)((IJvmObjectReferenceProvider)sc).Reference
+            .Jvm.CallStaticJavaMethod(
+                "org.apache.spark.api.python.PythonRDD",
+                "setupBroadcast",
+                _path);
+
+            SparkConf sparkConf = sc.GetConf();
+            bool encryptionEnabled = bool.Parse(
```

Comment:
Is this the right way to get this setting?   Pyspark calls [this method](https://github.com/apache/spark/blob/f378c7fba29368ca32142a3b7fc169dabe6cb37f/core/src/main/scala/org/apache/spark/api/python/PythonUtils.scala#L81)

Reply:
Yes I think that is the same spark API we are calling through our `SparkConf.Get()` API

Reply:
Although the config key is probably not going to change, wouldn't it be safer to call the method that will get us the correct key in case it does change in the future ?

Reply:
Ah ok sure, although I see this API name has [changed ](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/api/python/PythonUtils.scala#L81) in the latest Spark version.
So I will have to call different APIs based on spark version. Is that what we want to do @imback82

Reply:
Since we already know the API is changing, let's just use "spark.io.encryption" directly. Btw, let's expose `Get` in `SparkConf` so that we don't need to pass the default value.

Reply:
Looks like using `Get` (without the default value) throws a `NoSuchElementException` from Spark. I think it means if it is not set in the SparkConf, it will not return the Spark default.

Reply:
Right, but this is always set no?

Reply:
That would have been my guess based on the [documentation](https://spark.apache.org/docs/latest/security.html), but it looks like `getAll` returns default config but not `get`, as I see that is failing with the exception.

Reply:
Actually, I take this back and agree with @suhsteve. Instead of documenting this, let's just call `PythonUtils.getEncryptionEnabled`. When we run against Spark 3.0, the tests will fail anyway so it wouldn't go unnoticed. And now that it's super easy to get the spark version, I don't see an issue. Sorry for the confusion.

Reply:
@imback82 looks like `PythonUtils.getEncryptionEnabled` was introduced from Spark 2.3.3 and did not exist before that. Do we want to use our previous approach before that and call `getEncryptionEnabled` from 2.3.3 - 2.4.5 and `isEncryptionEnabled` 3.0.0 onwards?

Reply:
ok, then nothing much we can do except for sticking with spark conf. :(

Reply:
Can you now document?